com_migratetojoomla/admin/src/controller/MainController.php

```diff
@@ -0,0 +1,278 @@
+ */
+
+namespace Joomla\Component\MigrateToJoomla\Administrator\Controller;
+
+use Joomla\Database\DatabaseFactory;
+use Joomla\CMS\MVC\Controller\FormController;
+use Joomla\CMS\Factory;
+use Joomla\CMS\Language\Text;
+use Joomla\CMS\Router\Route;
+use Joomla\CMS\Versioning\VersionableControllerTrait;
+use Joomla\Component\MigrateToJoomla\Administrator\Helper\MainHelper;
+use Joomla\Component\MigrateToJoomla\Administrator\Helper\HttpHelper;
+use Joomla\Component\MigrateToJoomla\Administrator\Helper\FilesystemHelper;
+use Joomla\Component\MigrateToJoomla\Administrator\Helper\FtpHelper;
+use Joomla\CMS\Filesystem\Folder;
+
+// phpcs:disable PSR1.Files.SideEffects
+\defined('_JEXEC') or die;
+// phpcs:enable PSR1.Files.SideEffects
+
+/**
+ * main controller class.
+ *
+ * @since  1.0
+ */
+class MainController extends FormController
+{
+    use VersionableControllerTrait;
+
+    /**
+     * @var object  Media Download object
+     * 
+     * @since 1.0
+     */
+    public  $mediaDownloadManager;
+
+    /**
+     * @var object Database parameters object
+     * 
+     * @since 1.0
+     */
+    public $options;
+
+    /**
+     * @var object Database object
+     * 
+     * @since 1.0
+     */
+    public $db;
+
+    /**
+     * The prefix to use with controller messages.
+     *
+     * @var    string
+     * @since  
+     */
+    protected $text_prefix = 'COM_MIGRATETOJOOMLA_MAIN';
+
+    /**
+     * Method to check media connection.
+     * 
+     * @since 1.0
+     */
+    public function checkMediaConnection()
+    {
+        $this->checkToken();
+
+        $app   = Factory::getApplication();
+
+        $data  = $this->input->post->get('jform', array(), 'array');
+
+        $this->testMediaConnection($data);
+        // Store data in session
+        $app->setUserState('com_migratetojoomla.main', $data);
+
+        // redirect in all case
+        $this->setRedirect(Route::_('index.php?option=com_migratetojoomla', false));
+    }
+
+    /**
+     * Controller funciton to check Database connection
+     * 
+     * @since 1.0
+     */
+    public function checkDatabaseConnection()
+    {
+        $this->checkToken();
+
+        $app   = Factory::getApplication();
+        $data  = $this->input->post->get('jform', array(), 'array');
+        $isconnected = MainHelper::testDatabaseConnection($data);
+
+        if ($isconnected) {
+            $app->enqueueMessage(TEXT::_('COM_MIGRATETOJOOMLA_DATABASE_CONNECTION_SUCCESSFULLY'), 'success');
+        } else {
+            $app->enqueueMessage(TEXT::_('COM_MIGRATETOJOOMLA_DATABASE_CONNECTION_UNSUCCESSFULLY'), 'danger');
+        }
+        // Store data in session
+        $app->setUserState('com_migratetojoomla.main', $data);
+
+        // redirect in all case
+        $this->setRedirect(Route::_('index.php?option=com_migratetojoomla', false));
+    }
+
+    /**
+     * Method to Download 
+     * 
+     * @since  1.0
+     */
+    public function download()
```

Comment:
This will break with any bigger webpage.

The migration has to be done in many small steps, copying only a few files and then the next (ajax).

Reply:
Yes I have done migration in small small step. this method is starting point